[HUDI-845] Added locking capability to allow multiple writers

n3nash opened the pull request:
1. Added LockProvider API for pluggable lock methologies
2. Added Resolution Strategy API to allow for pluggable conflict resolution

**This diff -> https://github.com/apache/hudi/pull/2359 is a pre-requisite for landing this PR.**

## *Tips*
- *Thank you very much for contributing to Apache Hudi.*
- *Please review https://hudi.apache.org/contributing.html before opening a pull request.*

## What is the purpose of the pull request

*(For example: This pull request adds quick-start document.)*

## Brief change log

*(for example:)*
  - *Modify AnnotationLocation checkstyle rule in checkstyle.xml*

## Verify this pull request

*(Please pick either of the following options)*

This pull request is a trivial rework / code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

*(example:)*

  - *Added integration tests for end-to-end.*
  - *Added HoodieClientWriteTest to verify the change.*
  - *Manually verified the change by running a job locally.*

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieClient.java:
```diff
@@ -48,6 +50,7 @@
   protected final transient Configuration hadoopConf;
   protected final HoodieWriteConfig config;
   protected final String basePath;
+  protected AtomicReference<Option<HoodieInstant>> latestWriteInstantCompletedBeforeWriter = new AtomicReference<>();
```
AtomicReference to handle multiple threads using the same client object? +1

vinothchandar replied:
Since this is specific to writing, can we put this in the AbstractHoodieWriteClient subclass?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -203,6 +208,45 @@ public boolean commitStats(String instantTime, List<HoodieWriteStat> stats, Opti
+
+    HoodieCommitMetadata metadata = CommitUtils.buildMetadata(stats, partitionToReplaceFileIds, extraMetadata,
+        operationType, config.getSchema(), commitActionType);
+    if (config.isMultiWriterEnabled()) {
```
use the concurrency mode config introduced by the other PR and use that instead to fence this block?

n3nash replied:
That's the plan, once it's landed will change this, for now, this config is just a place holder.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -203,6 +208,45 @@ public boolean commitStats(String instantTime, List<HoodieWriteStat> stats, Opti
+        // TODO : Get timeout and set the timeout
+        boolean acquired = lockProvider.tryLock();
+        if (!acquired) {
+          throw new HoodieException("Unable to acquire lock");
```
some more context on basePath, inner exception if any ?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -203,6 +208,45 @@ public boolean commitStats(String instantTime, List<HoodieWriteStat> stats, Opti
+          throw new HoodieException("Unable to acquire lock");
+        }
+        LOG.info("Acquired lock for instant time " + instantTime);
+        // TODO : Move the following to a "critical section"
```
is TODO still valid?

n3nash replied:
Not anymore

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -203,6 +208,45 @@ public boolean commitStats(String instantTime, List<HoodieWriteStat> stats, Opti
+        try {
+          metadata = resolveWriteConflictIfAny(table, instantTime, metadata);
+          return commitStats(instantTime, stats, extraMetadata, commitActionType, Collections.emptyMap(), metadata);
+        } catch (Exception e) {
```
We need a new Exception class here. `HoodieWriteConflictException` or sth

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -203,6 +208,45 @@ public boolean commitStats(String instantTime, List<HoodieWriteStat> stats, Opti
+          return commitStats(instantTime, stats, extraMetadata, commitActionType, Collections.emptyMap(), metadata);
+        } catch (Exception e) {
+          // if strategy throws exception, first release lock
+          lockProvider.unlock();
```
why unlock here again, when you do it in finally

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -821,6 +865,38 @@ protected void finalizeWrite(HoodieTable<T, I, K, O> table, String instantTime,
 
+  private HoodieCommitMetadata resolveWriteConflictIfAny(HoodieTable<T, I, K, O> table, final String instantTime,
+                                            HoodieCommitMetadata thisCommitMetadata) {
+    Long currentTimeInSecs = System.currentTimeMillis() / 1000;
```
I was expecting to use the atomic reference above. and just pull all the writes that happened after the `latestWriteInstantCompletedBeforeWriter` ? Also lets not assume that time in seconds is what we will use here.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -203,6 +208,45 @@ public boolean commitStats(String instantTime, List<HoodieWriteStat> stats, Opti
+    if (config.isMultiWriterEnabled()) {
+      // get strategy and lock type
+      LockConfiguration lockConfiguration = new LockConfiguration(config.getProps());
+      LockProvider lockProvider = (LockProvider) ReflectionUtils.loadClass(config.getLockProviderClass(),
```
can we move the reflection part to the config object?

n3nash replied:
Already have it that way for `ConflictResolutionStrategy`, but this reflection requires parameters hence kept it here. If we moved this to config object, we have to make the getLockProviderClass parameterized which will be a problem for clients to get just the class name. Other option is to have 2 overloaded getters. Let me know what you think.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -821,6 +865,38 @@ protected void finalizeWrite(HoodieTable<T, I, K, O> table, String instantTime,
+        .findInstantsInRange(lastInstantTimestamp, String.valueOf(currentTimeInSecs))
+        .getInstants();
+
+    LOG.info("Current eligible instants during write oHoodieWriteConfigf instant " + instantTime + " = "
```
typos

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -821,6 +865,38 @@ protected void finalizeWrite(HoodieTable<T, I, K, O> table, String instantTime,
+    });
+    HoodieCommitMetadata newMetadataAfterConflictResolution = thisCommitMetadata;
+    if (hasConflict) {
+      newMetadataAfterConflictResolution = resolutionStrategy.resolveConflict(config, hadoopConf);
```
we should probably have an API to resolveConflict(instant1, instant2) ? i.e like our preCombine method? that will give resolutionStrategy implementors to just reason between two instants and whether they conflict or not, and if they do, then munge the commit metadata somehow.

n3nash replied:
Done

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/lock/ConcurrentFileWritesConflictResolutionStrategy.java:
```diff
@@ -0,0 +1,54 @@
+
+  @Override
+  public HoodieCommitMetadata resolveConflict(HoodieWriteConfig config, Configuration configuration) {
+    throw new UnsupportedOperationException("Cannot resolve conflicts for overlapping writes");
```
throw a special purpose exception here please

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/lock/ConcurrentFileWritesConflictResolutionStrategy.java:
```diff
@@ -0,0 +1,54 @@
+    Set<String> intersection = new HashSet<>(fileIdsSetForFirstInstant);
+    intersection.retainAll(fileIdsSetForSecondInstant);
+    if (!intersection.isEmpty()) {
+      LOG.error("Found conflicting writes " + intersection);
```
more information on these logs

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/lock/ConflictResolutionStrategy.java:
```diff
@@ -0,0 +1,35 @@
+ */
+public interface ConflictResolutionStrategy {
+
+  boolean hasConflict(HoodieCommitMetadata c1, HoodieCommitMetadata c2);
```
please add APIMaturity annotations and javadocs.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/lock/FileSystemBasedLockProvider.java:
```diff
@@ -0,0 +1,97 @@
+    try {
+      this.lockConfiguration = lockConfiguration;
+      this.lockPath = lockConfiguration.getConfig().getString(FILESYSTEM_LOCK_PATH_PROP);
+      this.fs = FileSystem.get(configuration);
```
please use FSUtils.getFS() always ! to get the filesystem object

n3nash replied:
Moved this to test scope

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/lock/FileSystemBasedLockProvider.java:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  @Override
+  public void unlock() {
```
whats the guarantee that `unlock()` will fail for thread B, if thread A had actually created the lock file. Don't think this will work. We should remove this implementation, if we cannot get it right IMO

vinothchandar replied:
which is indeed hard. I dont its possible on cloud stores.

n3nash replied:
This was just implemented for my testing (like mentioned in the comments), I've moved it to test scope.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/lock/ZookeeperBasedLockProvider.java:
```diff
@@ -0,0 +1,135 @@
+      }
+      lock.get().release();
+      lock.set(null);
+      LOG.info("Released lock");
```
more context?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieLockConfig.java:
```diff
@@ -0,0 +1,145 @@
+public class HoodieLockConfig extends DefaultHoodieConfig {
+
+  // Pluggable type of lock provider
+  public static final String HOODIE_LOCK_PROVIDER_CLASS_PROP = HOODIE_LOCK_PREFIX + "provider";
```
lets drop the `HOODIE_` prefix from the properties, per how we have named the configs thus far

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java:
```diff
@@ -122,6 +123,10 @@
   private static final String DEFAULT_MERGE_DATA_VALIDATION_CHECK_ENABLED = "false";
 
+  // Enable multi writer support
+  private static final String HOODIE_TABLE_MULTIWRITER_ENABLED_PROP = "hoodie.table.multiwriter.enabled";
```
repeating my earlier comment; lets just use the concurrency mode config from the other PR .

n3nash replied:
Responded earlier

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java:
```diff
@@ -385,6 +386,10 @@ protected void completeClustering(HoodieReplaceCommitMetadata metadata, JavaRDD<
     } else {
       writeTimer = metrics.getDeltaCommitCtx();
     }
+    latestWriteInstantCompletedBeforeWriter.set(table.getMetaClient().getActiveTimeline().getAllCommitsTimeline()
```
so, I was expecting we will do some kind of atomic swap, not just a set. if you just want to set, even a volatile variable is fine right?

n3nash replied:
That was the intention, missed this, fixed now

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/cluster/SparkExecuteClusteringCommitActionExecutor.java:
```diff
@@ -118,7 +118,7 @@ public SparkExecuteClusteringCommitActionExecutor(HoodieEngineContext context,
       return ((ClusteringExecutionStrategy<T, JavaRDD<HoodieRecord<? extends HoodieRecordPayload>>, JavaRDD<HoodieKey>, JavaRDD<WriteStatus>>)
           ReflectionUtils.loadClass(config.getClusteringExecutionStrategyClass(), table, context, config))
-          .performClustering(inputRecords, clusteringGroup.getNumOutputFileGroups(), instantTime, strategyParams, readerSchema);
+          .performClustering(inputRecords, 0, instantTime, strategyParams, readerSchema);
```
why is this relevant here?

n3nash replied:
The build was failing and not able to find this because I probably hadn't built the avro classes, so these changes were just a workaround, they have been reverted now.

vinothchandar commented on hudi-client/hudi-spark-client/src/main/scala/org/apache/hudi/AvroConversionHelper.scala:
```diff
@@ -314,7 +313,7 @@ object AvroConversionHelper {
             val sourceArray = item.asInstanceOf[Seq[Any]]
             val sourceArraySize = sourceArray.size
-            val targetList = new util.ArrayList[Any](sourceArraySize)
+            val targetList = new java.util.ArrayList[Any](sourceArraySize)
```
back these changes out?

n3nash replied:
Reverted these changes, but for some reason my intellij complains and cannot find util.Arraylist etc

vinothchandar commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/client/TestHoodieClientOnCopyOnWriteStorage.java:
```diff
@@ -216,6 +222,71 @@ public void testDeduplicationOnUpsert() throws Exception {
   }
 
+  @Test
+  public void testMultiWriter() throws Exception {
```
lets add a new test around the multi writing please. it should ideally be agnostic of COW/MOR. and we can just test MOR.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/config/LockConfiguration.java:
```diff
@@ -0,0 +1,62 @@
+public class LockConfiguration {
+
+  public static final String HOODIE_LOCK_PREFIX = "hoodie.writer.lock.";
+  public static final String HOODIE_LOCK_ACQUIRE_RETRY_WAIT_TIME_IN_MILLIS_PROP = HOODIE_LOCK_PREFIX + "wait_time_ms";
```
Drop the `HOODIE_` ?

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/config/LockConfiguration.java:
```diff
@@ -0,0 +1,62 @@
+ * Configuration for managing locks. Since this configuration needs to be shared with HiveMetaStore based lock,
+ * which is in a different package than other lock providers, we use this as a data transfer object in hoodie-common
+ */
+public class LockConfiguration {
```
whats the reason to have this in `hudi-common`? locking is only used by writing correct.

n3nash replied:
So we need to pass the same configs to `HiveMetastoreBasedLockProvider` which is in hudi-hive-sync as well as `ZookeeperBasedLockProvider` which is in hudi-client. This object helps to keep the config reading from the same class.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/lock/LockProvider.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * Pluggable lock implementations using this provider class.
+ */
+public abstract class LockProvider implements Lock, AutoCloseable {
```
can this be an interface?

n3nash replied:
Any specific reason you want this to be an interface ?

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/lock/LockProvider.java:
```diff
@@ -0,0 +1,66 @@
+  }
+
+  @Override
+  public final boolean tryLock(long time, TimeUnit unit) {
```
we should probably support this right. even of the `lock()` , since this has a timeout param?

n3nash replied:
Had implemented this differently earlier, changed some of the contracts, take a look again please

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/HoodieCommitMetadata.java:
```diff
@@ -109,6 +108,17 @@ public void setCompacted(Boolean compacted) {
     return filePaths;
   }
 
+  public HashMap<String, String> getFileIdWithoutSuffixAndRelativePaths() {
```
can this sit somewhere else? in a helper method?

n3nash replied:
done

vinothchandar commented on hudi-integ-test/src/main/java/org/apache/hudi/integ/testsuite/generator/GenericRecordFullPayloadGenerator.java:
```diff
@@ -284,7 +284,7 @@ private Object typeConvert(Schema.Field field) {
     // TODO: Need to implement valid data generation for fixed type
     GenericFixed genericFixed = new GenericData.Fixed(localSchema);
-    switch (localSchema.getLogicalType().getName()) {
+    switch (localSchema.getType().getName()) {
```
why this change?

n3nash replied:
Same, my intellij doesn't work, reverted

vinothchandar reviewed (commented):
some high level comments. High level, looks promising to me.

n3nash commented:
@vinothchandar Addressed most of the code comments replied to some which needs clarification/resolution. 2 high level items are missing 1. Adding a new test class for multi-writer 2. Testing with test-suite / real job.

n3nash commented:
@vinothchandar This PR is ready, concurrency control added for writers and table services, documentation to follow after this PR is merged.

vinothchandar commented:
@n3nash Beginning the final review of this. Should we still land #2359 first?

vinothchandar commented:
@n3nash can you please rebase and repush

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -859,6 +973,68 @@ protected void finalizeWrite(HoodieTable<T, I, K, O> table, String instantTime,
+    return thisOperation.getCommitMetadata();
+  }
+
+  private boolean executeCriticalSection(String instantTime, List<HoodieWriteStat> stats, Option<Map<String, String>> extraMetadata,
```
same with this.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -907,11 +1083,20 @@ public void close() {
     // Calling this here releases any resources used by your index, so make sure to finish any related operations
     // before this point
     this.index.close();
+    // HiveMetastoreClient does not implement AutoCloseable. Additionally, we cannot call close() after unlock()
```
please move these comments into the actual implementation. and not in this file

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieCompactionConfig.java:
```diff
@@ -41,7 +41,8 @@
   public static final String ASYNC_CLEAN_PROP = "hoodie.clean.async";
-
+  // Turn on inline cleaning
+  public static final String INLINE_CLEAN_PROP = "hoodie.clean.inline";
```
we should check somewhere that user cannot turn on both async and inline? , if we are adding an explicit config.

n3nash replied:
Right now, this config is harmless, it just assumes the role of autoClean.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java:
```diff
@@ -389,6 +397,10 @@ public boolean isAsyncClean() {
     return Boolean.parseBoolean(props.getProperty(HoodieCompactionConfig.ASYNC_CLEAN_PROP));
   }
 
+  public boolean isInlineCleaning() {
```
this does not read easily. rename `inlineCleaningEnabled` or `shouldCleanInline()` or something like tht? (same wherever applicable)

n3nash replied:
I kept the same naming convention as before for `isInlineCompaction` etc. I refactored all of them to `inline<..>Enabled`

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java:
```diff
@@ -923,6 +935,39 @@ public int getMetadataCleanerCommitsRetained() {
+    return WriteConcurrencyMode.fromValue(props.getProperty(WRITE_CONCURRENCY_MODE_PROP));
+  }
+
+  public Boolean isInlineTableServiceEnabled() {
```
same. rename `inlineTableServices()`?

n3nash replied:
done

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -88,7 +88,7 @@
   protected final transient HoodieEngineContext engineContext;
 
-  protected HoodieBackedTableMetadataWriter(Configuration hadoopConf, HoodieWriteConfig writeConfig, HoodieEngineContext engineContext) {
+  protected HoodieBackedTableMetadataWriter(Configuration hadoopConf, HoodieWriteConfig writeConfig, HoodieEngineContext engineContext, boolean bootstrapIfNeeded) {
```
rename to `initIfNeeded` consistently

n3nash replied:
refactored to `bootstrapMetadata`

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -230,13 +235,18 @@ protected void initTableMetadata() {
-      // Initialize for the first time by listing partitions and files directly from the file system
-      bootstrapFromFilesystem(engineContext, datasetMetaClient);
-      metrics.ifPresent(m -> m.updateMetrics(HoodieMetadataMetrics.INITIALIZE_STR, timer.endTimer()));
+  protected void bootstrapIfNeeded(HoodieEngineContext engineContext, HoodieTableMetaClient datasetMetaClient) {
```
we can also rename this method if needed

n3nash replied:
refactored to `bootstrapMetadata`

vinothchandar reviewed (commented):
Can we move the scheduling of compaction and clustering into the writeClient itself, protected in the critical section. 

I can review the lock implementations here more line-by-line. but this is the main gap that I saw

n3nash commented on hudi-common/src/main/java/org/apache/hudi/common/model/WriteConcurrencyMode.java:
```diff
@@ -0,0 +1,76 @@
+ * Different concurrency modes for write operations.
+ */
+public enum WriteConcurrencyMode {
+  NO_WRITER("no_writer"),
```
Old code, removed.

n3nash commented on hudi-common/src/main/java/org/apache/hudi/common/model/TableService.java:
```diff
@@ -0,0 +1,62 @@
+ * Supported runtime table services.
+ */
+public enum TableService {
+  COMPACT("compact"),
```
It should be, but for each of these table services, we now follow the steps of `schedule` -> `inflight` -> `complete`. Archival doesn't do that right now, so I haven't added it here. I have refactored the `clean` actions to do this.

n3nash commented on hudi-common/src/main/java/org/apache/hudi/common/model/TableService.java:
```diff
@@ -0,0 +1,62 @@
+  COMPACT("compact"),
+  CLUSTER("cluster"),
+  CLEAN("clean");
+  private final String value;
```
removed

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -220,7 +253,7 @@ void emitCommitMetrics(String instantTime, HoodieCommitMetadata metadata, String
   }
 
-  protected void syncTableMetadata() {
+  protected void syncTableMetadata(boolean bootstrapIfNeeded) {
```
`init` also has a different meaning inside the `HoodieBackedTableMetadataWriter` where init just means initializing the table metadata. How about just `bootstrapMetadata` ?

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -403,30 +439,32 @@ protected void postCommit(HoodieTable<T, I, K, O> table, HoodieCommitMetadata me
-      } else {
-        metadata.addMetadata(HoodieClusteringConfig.INLINE_CLUSTERING_PROP, "false");
-      }
+        // Do an inline clustering if enabled
```
Intellij says ok, not sure why it's appearing this way here

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -599,6 +637,7 @@ public HoodieRestoreMetadata restoreToInstant(final String instantTime) throws H
   public HoodieCleanMetadata clean(String cleanInstantTime) throws HoodieIOException {
     LOG.info("Cleaner started");
     final Timer.Context timerContext = metrics.getCleanCtx();
+    scheduleCleaningAtInstant(cleanInstantTime, Option.empty());
```
I refactored the code to ensure we follow `schedule` -> `inflight` -> `complete` for clean actions as well.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -403,30 +439,32 @@ protected void postCommit(HoodieTable<T, I, K, O> table, HoodieCommitMetadata me
-      } else {
-        metadata.addMetadata(HoodieCompactionConfig.INLINE_COMPACT_PROP, "false");
-      }
+      if (config.isInlineTableServiceEnabled()) {
```
I just wrapped all table services together as inline or async. The scheduling is protected via lock in inline as well as async.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -98,6 +112,7 @@
   private transient HoodieWriteCommitCallback commitCallback;
   private transient AsyncCleanerService asyncCleanerService;
   protected final boolean rollbackPending;
+  protected AtomicReference<Option<HoodieInstant>> latestCompletedWriteInstant = new AtomicReference<>();
```
Created issue -> https://issues.apache.org/jira/browse/HUDI-1577

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -239,6 +272,9 @@ public void bootstrap(Option<Map<String, String>> extraMetadata) {
     if (rollbackPending) {
       rollBackInflightBootstrap();
     }
+    if (config.getWriteConcurrencyMode().isMultiWriter()) {
```
Right now, we don't have multi-writer so people might remain cautious. Once we support multiwriter, it's easy to set one config and run bootstrap/incremental, guarding against such behaviors. 
Right now, do we have a way to enforce users to downgrade to a single writer ?

n3nash commented:
@vinothchandar Addressed your comments and replied to other ones, PTAL. I've also refactored the code to make the flow simpler.

n3nash commented:
@vinothchandar Code is ready for review.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/config/LockConfiguration.java:
```diff
@@ -0,0 +1,69 @@
+  public static final String HIVE_TABLE_NAME_PROP = HIVE_METASTORE_LOCK_PROPERTY_PREFIX + "table";
+  // Zookeeper configs for zk based locks
+  public static final String ZOOKEEPER_BASED_LOCK_PROPERTY_PREFIX = LOCK_PREFIX + "zookeeper.";
+  public static final String ZK_BASE_PATH_PROP = ZOOKEEPER_BASED_LOCK_PROPERTY_PREFIX + "zk_base_path";
```
I think the common term is . zookeeper chroot?

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/lock/LockProvider.java:
```diff
@@ -0,0 +1,86 @@
+/**
+ * Pluggable lock implementations using this provider class.
+ */
+public abstract class LockProvider<T> implements Lock, AutoCloseable {
```
can this be an interface?

vinothchandar replied:
if there is no shared code here, we should go for an interface vs an abstract class

vinothchandar replied:
We can always to interface and default methods, right?

n3nash replied:
I want to make some methods final which we cannot do in interfaces. But it's fine, I have changed it to interface, I don't have a very strong preference.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/HoodieCommonMetadata.java:
```diff
@@ -0,0 +1,33 @@
+
+import org.apache.hudi.avro.model.HoodieArchivedMetaEntry;
+
+public class HoodieCommonMetadata {
```
rename: HoodieMetadataWrapper

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/TableService.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Supported runtime table services.
+ */
+public enum TableService {
```
rename: TableServiceType

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/WriteConcurrencyMode.java:
```diff
@@ -0,0 +1,66 @@
+    }
+  }
+
+  public boolean supportsOptimisticConcurrencyControl() {
```
rename: isOptimistic... ?

vinothchandar replied:
I was thinking from perspective of, you are just doing a direct comparison. thus hte `isXXX` naming. but I see your view. lets keep it as is

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/WriteOperationType.java:
```diff
@@ -82,6 +84,10 @@ public static WriteOperationType fromValue(String value) {
         return CLUSTER;
+      case "compact":
+        return COMPACT;
+      case "unknown":
```
what is this really?  how can there be an write that is unknown?

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/util/CommitUtils.java:
```diff
@@ -93,4 +95,28 @@ private static HoodieCommitMetadata buildMetadataFromStats(List<HoodieWriteStat>
     return commitMetadata;
   }
+
+  public static HashMap<String, String> getFileIdWithoutSuffixAndRelativePaths(Map<String, List<org.apache.hudi.avro.model.HoodieWriteStat>>
```
lets unit tests these?

vinothchandar commented on hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/client/HoodieFlinkWriteClient.java:
```diff
@@ -313,6 +320,16 @@ public void cleanHandles() {
+    // Create a Hoodie table after startTxn which encapsulated the commits and files visible.
+    // Important to create this after the lock to ensure latest commits show up in the timeline without need for reload
+    HoodieTable table = createTable(config, hadoopConf);
+    // TODO : Metadata Writer is not supported for Flink
```
revisit todo?

vinothchandar commented on hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/table/HoodieFlinkCopyOnWriteTable.java:
```diff
@@ -265,6 +266,20 @@ public void rollbackBootstrap(HoodieEngineContext context, String instantTime) {
+   * @return
+   */
+  @Override
+  public Option<HoodieCleanerPlan> scheduleCleaning(HoodieEngineContext context, String instantTime, Option<Map<String, String>> extraMetadata) {
```
this does not seem ok to do?

n3nash replied:
Yeah, had a TODO on this, addressed it now.

vinothchandar commented on hudi-client/hudi-java-client/src/main/java/org/apache/hudi/client/HoodieJavaWriteClient.java:
```diff
@@ -224,6 +231,16 @@ protected void completeCompaction(HoodieCommitMetadata metadata,
+    // Create a Hoodie table after startTxn which encapsulated the commits and files visible.
+    // Important to create this after the lock to ensure latest commits show up in the timeline without need for reload
+    HoodieTable table = createTable(config, hadoopConf);
+    // TODO : Metadata Writer is not supported for Java
```
revisit comment.

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java:
```diff
@@ -380,11 +391,40 @@ protected void completeClustering(HoodieReplaceCommitMetadata metadata, JavaRDD<
-    new SparkUpgradeDowngrade(metaClient, config, context).run(metaClient, HoodieTableVersion.current(), config, context, instantTime);
-    return getTableAndInitCtx(metaClient, operationType);
+    if (HoodieTableVersion.current() != metaClient.getTableConfig().getTableVersion()) {
+      // TODO : Force clean up of all inflights, do this once pending rollback removal PR is landed
```
lets clean this up?

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java:
```diff
@@ -380,11 +391,40 @@ protected void completeClustering(HoodieReplaceCommitMetadata metadata, JavaRDD<
     HoodieTableMetaClient metaClient = createMetaClient(true);
-    new SparkUpgradeDowngrade(metaClient, config, context).run(metaClient, HoodieTableVersion.current(), config, context, instantTime);
-    return getTableAndInitCtx(metaClient, operationType);
+    if (HoodieTableVersion.current() != metaClient.getTableConfig().getTableVersion()) {
```
lets add a method `needsUpgradeOrDowngrade() ` to the upgradedowngrade class?

n3nash replied:
Added

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java:
```diff
@@ -380,11 +391,40 @@ protected void completeClustering(HoodieReplaceCommitMetadata metadata, JavaRDD<
+    if (HoodieTableVersion.current() != metaClient.getTableConfig().getTableVersion()) {
+      // TODO : Force clean up of all inflights, do this once pending rollback removal PR is landed
+      // this.rollbackFailedWrites();
+      this.txnManager.beginTransaction();
```
should we check concurrency mode before taking these locks?

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java:
```diff
@@ -380,11 +391,40 @@ protected void completeClustering(HoodieReplaceCommitMetadata metadata, JavaRDD<
   }
 
-  private HoodieTable<T, JavaRDD<HoodieRecord<T>>, JavaRDD<HoodieKey>, JavaRDD<WriteStatus>> getTableAndInitCtx(HoodieTableMetaClient metaClient, WriteOperationType operationType) {
+  // TODO : To enforce priority between table service and ingestion writer, use transactions here and invoke strategy
```
comment valid? What should we do about this issue? Can you elabortate?

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java:
```diff
@@ -401,13 +441,25 @@ protected void completeClustering(HoodieReplaceCommitMetadata metadata, JavaRDD<
     // Open up the metadata table again, for syncing
-    try (HoodieTableMetadataWriter writer = SparkHoodieBackedTableMetadataWriter.create(hadoopConf, config, context)) {
+    try {
+      HoodieTableMetadataWriter writer =
```
we lose auto closing by moving to  a regular try-catch. Why is this change needed?

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java:
```diff
@@ -401,13 +441,25 @@ protected void completeClustering(HoodieReplaceCommitMetadata metadata, JavaRDD<
+  protected void preCommit(String instantTime, HoodieCommitMetadata metadata) {
+    // Create a Hoodie table after startTxn which encapsulated the commits and files visible.
+    // Important to create this after the lock to ensure latest commits show up in the timeline without need for reload
+    HoodieTable table = createTable(config, hadoopConf);
```
should we recreate? again, we need to revisit this whole passing of metadatawriter to resolve conflict. Don't understand this part

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/bootstrap/SparkBootstrapCommitActionExecutor.java:
```diff
@@ -222,6 +225,17 @@ protected void commit(Option<Map<String, String>> extraMetadata, HoodieWriteMeta
   }
 
+  @Override
+  protected void syncTableMetadata() {
```
I think all this can be removed. from all action executors? can't we take the lock in post commit/write from the write client level?

vinothchandar replied:
can we please file a ticket for removing this auto commit stuff. it's kind of messy.

n3nash replied:
https://issues.apache.org/jira/browse/HUDI-1665

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieCompactionConfig.java:
```diff
@@ -43,7 +43,8 @@
   public static final String ASYNC_CLEAN_PROP = "hoodie.clean.async";
-
+  // Turn on inline cleaning
+  public static final String INLINE_CLEAN_PROP = "hoodie.clean.inline";
```
I was trying to keep the same concepts of `inline` for clustering, compact. The problem is there are autoClean & autoCommit but no `autoCompact` or `autoCluster` etc. Additionally, we used `inline` as the flag to toggle between inline & async for compact & cluster while we have chosen `async` as the flag for clean. 

I have removed `hoodie.clean.inline`. We can address these in another PR after we decide what convention to follow.

vinothchandar replied:
sure. the idea is to not proliferate, since it then becomes one more config to take care of.

n3nash reviewed (commented):
@vinothchandar Addressed review comments and responded to many of them. PTAL. 1 pending item to add some more tests for 2 classes. I will do another round of testing after incorporating final changes.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -30,16 +31,19 @@
 import org.apache.hudi.common.model.HoodieKey;
 import org.apache.hudi.common.model.HoodieRecordPayload;
 import org.apache.hudi.common.model.HoodieWriteStat;
+import org.apache.hudi.common.model.TableService;
```
If its an enum, its okay. I thought it had code. Still not a `model` per se, but we can do the moving in a different PR

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -359,10 +388,21 @@ public abstract O bulkInsertPreppedRecords(I preppedRecords, final String instan
+        .lastInstant());
+    LOG.info("Last Instant Cached by writer with instant " + instantTime + " is " + this.txnManager.getLastCompletedTransactionOwner());
+    this.txnManager.setTransactionOwner(Option.of(new HoodieInstant(State.INFLIGHT, metaClient.getCommitActionType(), instantTime)));
+    this.txnManager.beginTransaction();
```
Understood. but if some calls are needed prior to this, then either the object constructor should take them or the method should take them as arguments. Or we have a transaction builder or sorts. 
So I would say, if this is always the case, i.e the setters are needed, then we change the beginTransaction() signature for good, not just an overload.

n3nash replied:
Made changes accordingly

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -359,10 +388,21 @@ public abstract O bulkInsertPreppedRecords(I preppedRecords, final String instan
+    try {
+      syncTableMetadata();
+    } finally {
+      this.txnManager.endTransaction();
```
I was wondering about the following scenario. if the cleanup fails, I guess it throws an error in both cases also? if so, this is okay. gtg

n3nash replied:
Yes

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/lock/ConflictResolutionStrategy.java:
```diff
@@ -0,0 +1,64 @@
+   * @return
+   */
+  @PublicAPIMethod(maturity = ApiMaturityLevel.EVOLVING)
+  Option<HoodieCommitMetadata> resolveConflict(Option<HoodieBackedTableMetadataWriter> metadataWriter, HoodieTable table,
```
lets remove it. we can introduce it as needed. This will simplify the implementation, as it stands now.

n3nash replied:
Removed

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/lock/HoodieCommitOperation.java:
```diff
@@ -0,0 +1,142 @@
+ * Since we interchange payload types between AVRO specific records and POJO's, this object serves as
+ * a common payload to manage these conversions.
+ */
+public class HoodieCommitOperation {
```
So the thing is , this has the common metadata, overloading "commit" is always confusing in our code base, since `commit` often refers to a specific action type. Given its used specifically, in the conflict resolution scenario, I think somethihg like `ConcurrentOperation` captures the intent. Its just concurrent to the current operation, the conflict resolution will determine if its actually a conflict like you mentioned.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/lock/LockManager.java:
```diff
@@ -0,0 +1,125 @@
+          Thread.sleep(waitTimeInMs);
+          retryCount++;
+        }
+      } catch (Exception e) {
```
I think. for the retries, we should handle interrupted exception and continue retrying. Thats what I was getting at.

n3nash replied:
added

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/lock/SimpleConcurrentFileWritesConflictResolutionStrategy.java:
```diff
@@ -0,0 +1,102 @@
+    Set<String> intersection = new HashSet<>(fileIdsSetForFirstInstant);
+    intersection.retainAll(fileIdsSetForSecondInstant);
+    if (!intersection.isEmpty()) {
+      LOG.error("Found conflicting writes between first operation = " + thisOperation
```
`WARN` should indicate abnormal execution. So may be or may be not. I still think INFO is the cleanest, since this code is supposed to handle the conflcting case. For debugging, users can always turn it on.

n3nash replied:
Okay, made it INFO

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/lock/ZookeeperBasedLockProvider.java:
```diff
@@ -0,0 +1,157 @@
+
+package org.apache.hudi.client.lock;
+
+import org.apache.curator.framework.CuratorFramework;
```
In the bundles, we explicitly white list dependencies. So not sure how transitive dependencies would have been picked up.  How are you testing all this - not using the utilities/spark bundles? 

Standard practice for dependencies that are actually used in the projects code, is to explicitly deal with the dependency. We should not depend on hbase-server transitiively bringing it in.

vinothchandar replied:
```
vmacs:target vs$ jar tf hudi-spark-bundle_2.11-0.8.0-SNAPSHOT.jar | grep curator | wc -l
       0
vmacs:target vs$ jar tf hudi-spark-bundle_2.11-0.8.0-SNAPSHOT.jar | grep hbase | wc -l
    4734
vmacs:target vs$ 
```

Don't see curator, in fact.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/SimpleConcurrentFileWritesConflictResolutionStrategy.java:
```diff
@@ -0,0 +1,105 @@
+    // we handle this during conflict resolution and not treat the commit from compaction operation
+    // as a regular commit. Regular commit, deltacommits and replace are candidates for conflict
+    // replace is used for a) clustering without update support b) insert_overwrite both of which are
+    // candidates for conflict. We need to add CLUSTER here once it supports concurrent updates
```
lets file a JIRA for clustering/concurrent updates.

n3nash replied:
There is already one -> https://issues.apache.org/jira/browse/HUDI-1042

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/SimpleConcurrentFileWritesConflictResolutionStrategy.java:
```diff
@@ -0,0 +1,105 @@
+    // replace is used for a) clustering without update support b) insert_overwrite both of which are
+    // candidates for conflict. We need to add CLUSTER here once it supports concurrent updates
+    // like COMPACT.
+    if (otherOperation.getOperationType() == WriteOperationType.COMPACT
```
what about async clustering? would it raise an exception? may be it wont today, given its always picks non conflicting files alreayd?

vinothchandar replied:
I just want to ensure that async clustering can happen, as long as the the files don't overlap.

n3nash replied:
Yes, they will. I have a test case for it as well.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/lock/SimpleConcurrentFileWritesConflictResolutionStrategy.java:
```diff
@@ -0,0 +1,102 @@
+    // NOTE that any commits from table services such as compaction, clustering or cleaning since the
+    // overlapping of files is handled using MVCC. Since compaction is eventually written as commit, we need to ensure
+    // we handle this during conflict resolution and not treat the commit from compaction operation as a regular commit.
+    if (otherOperation.getOperationType() == WriteOperationType.UNKNOWN
```
See newer review, for followup

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/ZookeeperBasedLockProvider.java:
```diff
@@ -0,0 +1,161 @@
+  }
+
+  private void acquireLock(long time, TimeUnit unit) throws Exception {
+    ValidationUtils.checkArgument(this.lock == null, LockState.ALREADY_ACQUIRED.name());
```
a better validation message, than just the enum name?  (applicable for any such usage)

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/ZookeeperBasedLockProvider.java:
```diff
@@ -0,0 +1,161 @@
+        + this.lockConfiguration.getConfig().getString(ZK_LOCK_KEY_PROP));
+    boolean acquired = newLock.acquire(time, unit);
+    if (!acquired) {
+      throw new HoodieLockException(generateLogStatement(LockState.FAILED_TO_ACQUIRE, generateLogSuffixString()));
```
this will lead to a LockException being wrapped inside another LockException in tryLock()?

vinothchandar replied:
Can we handle this specially and just rethrow the lock exception without the wrapping.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/lock/ZookeeperBasedLockProvider.java:
```diff
@@ -0,0 +1,157 @@
+      acquireLock(time, unit);
+      LOG.info(generateLogStatement(LockState.ACQUIRED, generateLogSuffixString()));
+    } catch (Exception e) {
+      throw new HoodieLockException(generateLogStatement(LockState.FAILED_TO_ACQUIRE, generateLogSuffixString()), e);
```
I think the actual code you have know handles this.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieLockConfig.java:
```diff
@@ -0,0 +1,191 @@
+import java.io.IOException;
+import java.util.Properties;
+
+import static org.apache.hudi.common.config.LockConfiguration.DEFAULT_ACQUIRE_LOCK_WAIT_TIMEOUT_MS;
```
But, could you explain why even `HiveMetastoreLockProvider` that has to be in hudi-common? That was my main point. We will be needlessly increasing the weight of hudi-common, which is picked up by hudi-hadoop-mr for e.g.

n3nash replied:
`HiveMetastoreLockProvider` has always been in `hudi-hive-sync` package.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/lock/LockProvider.java:
```diff
@@ -0,0 +1,86 @@
+  @Override
+  public void close() {
+    try {
+      close();
```
is this a recursive call? won't it recurse infinitely?

n3nash replied:
Intended to super.close()

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/lock/LockProvider.java:
```diff
@@ -0,0 +1,86 @@
+  @Override
+  public boolean tryLock(long time, TimeUnit unit) {
+    try {
+      return tryLock(time, unit);
```
same here. infinite loop?

n3nash replied:
same as above

vinothchandar reviewed (commented):
Marked as many as I could as resolved from the previous review. New comments here and some comments left on unresolved comments from the previous one. 

We definitely need to fix the bundles and test, before we can land this.

n3nash commented:
@vinothchandar Addressed your feedback comments. Some more things to note:

1. Added explicit dependency on `curator` to control the versions, earlier brought in through transitive dependency. NOTE that the latest stable version for curator is 4.X but we have to use curator version 2.X since 4.X is not backwards compatible with Zookeeper 2.x and 3.X while curator 2.X is backwards and forward compatible. The latest stable version of curator brings in many new features which are not applicable to us, so it should be OK to use 2.X
2. Added Spark Datasource based nodes to HoodieTestSuite and added a test case to run through optimistic locking enabled with Spark datasource. NOTE that we can only test using `ZookeeperBasedLockProvider` for tests and cannot use `HiveMetastoreBasedLockProvider`. This is because embedded  hive metastore only allows one Hive client. When we use spark, spark's HiveExternalCatalog already creates an internal hive client, when hudi tries to create a new hive client through the `HiveMetastoreBasedLockProvider`, it throws underlying derby related exceptions. Some workarounds are to not use derby and are mentioned here -> https://issues.apache.org/jira/browse/HIVE-21302 but this requires a change to the underlying choice of DB for hive metastore itself. Hence, for now, we can only use ZK locks for in-memory unit tests. Ran it against production setup of HiveMetastore to validate this is only a unit test problem. 
3. Added TestSimpleConcurrentFileWritesConflictResolutionStrategy to ensure the right candidate streams are always used. 
4. Added another section to quickstart on how to enable optimistic locks. This PR -> https://github.com/apache/hudi/pull/2660 has the details. Validated this by turning on INFO logs on docker to confirm that locking works.

Some warnings ( I will add these to the docs when I send the PR for documentation)
1. `DeltaStreamer` cannot work with multi-writer. Currently we copy checkpoints from the previous commit to the next - this breaks when we perform multi-writing. 
2. Configs for retries and timeouts for locking can vary because of many factors : response times of file system APIs, number of commits to check conflict resolution against etc. We need to document that folks have to choose sane timeouts and retries for their use-case if defaults don't work. 
3. Incremental pull using TimelineAPI's will not work for clients. There will be a follow up PR for this in progress.

vinothchandar commented:
>DeltaStreamer cannot work with multi-writer. Currently we copy checkpoints from the previous commit to the next - this breaks when we perform multi-writing.

But we should be able to allow backfills using spark sql, which don't really mess with checkpoints? We cannot ignore this scenario. Can you look into how we can provide best practices

n3nash commented:
@vinothchandar It's possible to allow backfills using spark-sql but there are some corner cases. Consider the following:

1. Ingestion job running with commit c4 (checkpoint = c3)
2. Ingestion job finishes with commit c4 (checkpoint = c3)
3. Someone runs a spark-sql job to backfill some data in an older partition, commit c5. Since this spark-sql job (unlike deltastreamer) does not handle checkpoint copying from prev metadata to next, it would be the client's job to do this. 
4. If they fail to do this, deltastreamer next ingestion c6 will read no checkpoint from c5. 

I've made the following changes:

1) To make this manageable, I've added the following config : `hoodie.write.meta.key.prefixes`. One can set this config to ensure that during the critical section, if this config is set, it will copy over all the metadata for the keys that match with the prefix set in this config from the latest metadata to the current commit.
2) Added these multi-writer tests to `HoodieDeltaStreamer` as well. 

NOTE: The test may be failing because of some thread.sleep related code that I'm trying to remove. Will update tomorrow.

n3nash commented:
@vinothchandar Build succeeds locally and should pass on jenkins (will check tomorrow morning), ready for review.

vinothchandar reviewed (approved):
Just went over the feedback from my last review. 

1. have you tested the spark bundle end to end? I think so, based on #2660 . Just making sure. 
2. Can we also add the curator deps to flink and utilities bundles? 

Pending these, once CI is happy. We can land and do the follow ups.

n3nash commented:
@vinothchandar I've added curator to utilities bundle. In this PR, we don't support locking for flink and java clients so I have not added to the flink bundle. If you have another reason to add curator to flink bundles, let me know.

vinothchandar commented:
cc @yanghua @danny0405 @leesf to take note for flink/java support follow up.

n3nash commented:
Here is the issue for the follow up -> https://issues.apache.org/jira/browse/HUDI-1698

danny0405 commented:
> Here is the issue for the follow up -> https://issues.apache.org/jira/browse/HUDI-1698. I'll work with you guys offline @danny0405 @yanghua @leesf

Thanks for the notion, would take up the following work.
